### PR TITLE
[Dataset] Support dataset from a single dataframe/table.

### PR DESCRIPTION
**Pull request by sighingnow**
## Why are these changes needed?

Add support for creating dataset from a single dataframe and a single table.

It should be reasonable to support such behaviour, and, if users feed into just a single dataframe, the error message is quite hard to understand (as ray is trying to iterate over columns of the dataframe...):

```python
>>> df = pd.DataFrame({"x": [1,2,3]})
>>> ds = ray.data.from_pandas(df)
(pid=42071) df =  x <class 'str'>
---------------------------------------------------------------------------
RayTaskError(TypeError)                   Traceback (most recent call last)
<ipython-input-5-34d40ad42e52> in <module>
----> 1 ds = ray.data.from_pandas(df)

~/ray/python/ray/data/read_api.py in from_pandas(dfs, parallelism)
    474
    475     res = [df_to_block.remote(df) for df in dfs]
--> 476     blocks, metadata = zip(*res)
    477     return Dataset(BlockList(blocks, ray.get(list(metadata))))
    478

~/ray/python/ray/_private/client_mode_hook.py in wrapper(*args, **kwargs)
     80         if client_mode_should_convert():
     81             return getattr(ray, func.__name__)(*args, **kwargs)
---> 82         return func(*args, **kwargs)
     83
     84     return wrapper

~/ray/python/ray/worker.py in get(object_refs, timeout)
   1617                     worker.core_worker.dump_object_store_memory_usage()
   1618                 if isinstance(value, RayTaskError):
-> 1619                     raise value.as_instanceof_cause()
   1620                 else:
   1621                     raise value

RayTaskError(TypeError): ray::_df_to_block() (pid=42071, ip=30.25.96.43)
  File "/Users/linzhu.ht/ray/python/ray/data/read_api.py", line 516, in _df_to_block
    print('df = ', df, type(df))
  File "pyarrow/table.pxi", line 2228, in pyarrow.lib.table
TypeError: Expected pandas DataFrame, python dictionary or list of arrays

In [6]: 2021-08-30 17:56:28,638	ERROR worker.py:79 -- Unhandled error (suppress with RAY_IGNORE_UNHANDLED_ERRORS=1): ray::_df_to_block() (pid=42071, ip=30.25.96.43)
  File "/Users/linzhu.ht/ray/python/ray/data/read_api.py", line 516, in _df_to_block
    print('df = ', df, type(df))
  File "pyarrow/table.pxi", line 2228, in pyarrow.lib.table
TypeError: Expected pandas DataFrame, python dictionary or list of arrays
```

## Related issue number

None.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by clarkzinzow — COMMENTED** [review]
This looks good to me, thanks for making this change! Could you add a quick test case to the `from_pandas` and `from_arrow` tests, [here](https://github.com/ray-project/ray/blob/529508125d27c89d9afaa32036444128e21712c1/python/ray/data/tests/test_dataset.py#L388-L394) and [here](https://github.com/ray-project/ray/blob/529508125d27c89d9afaa32036444128e21712c1/python/ray/data/tests/test_dataset.py#L405-L414)? After that, this looks ready to merge.

**Comment by sighingnow** [comment]
> here

Thanks for reviewing. I have rebase to master and add test cases.

**Review by clarkzinzow — APPROVED** [review]
LGTM! Thanks for contributing this!

**Comment by clarkzinzow** [comment]
Ah @sighingnow looks like there's a related test failure:
```
def test_from_arrow(ray_start_regular_shared):
--
  | df1 = pd.DataFrame({"one": [1, 2, 3], "two": ["a", "b", "c"]})
  | df2 = pd.DataFrame({"one": [4, 5, 6], "two": ["e", "f", "g"]})
  | ds = ray.data.from_arrow([
  | ray.put(pa.Table.from_pandas(df1)),
  | ray.put(pa.Table.from_pandas(df2))
  | ])
  | values = [(r["one"], r["two"]) for r in ds.take(6)]
  | rows = [(r.one, r.two) for _, r in pd.concat([df1, df2]).iterrows()]
  | assert values == rows
  |  
  | # test from single pyarrow table
  | >       ds = ray.data.from_arrow(ray.put(pa.Table.from_pandas(df1)))
  |  
  | /ray/python/ray/data/tests/test_dataset.py:858:
  | _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
  |  
  | tables = ObjectRef(ffffffffffffffffffffffffffffffffffffffff010000003a000000)
  |  
  | @PublicAPI(stability="beta")
  | def from_arrow(tables: Union[ObjectRef["pyarrow.Table"], List[ObjectRef[
  | "pyarrow.Table"]]]) -> Dataset[ArrowRow]:
  | """Create a dataset from a set of Arrow tables.
  |  
  | Args:
  | dfs: A Ray object references to Arrow tables or a list of such
  | references.
  |  
  | Returns:
  | Dataset holding Arrow records from the tables.
  | """
  | import pyarrow as pa
  |  
  | if isinstance(tables, pa.Table):
  | tables = [tables]
  |  
  | get_metadata = cached_remote_fn(_get_metadata)
  | >       metadata = [get_metadata.remote(t) for t in tables]
  | E       TypeError: 'ray._raylet.ObjectRef' object is not iterable
```

File: python/ray/data/read_api.py [hunk 1]
```diff
@@ -513,15 +513,22 @@ def from_modin(df: "modin.DataFrame") -> Dataset[ArrowRow]:
...
-        dfs: A list of Ray object references to pandas dataframes.
+        dfs: A Ray object reference to pandas dataframes or a list of
+            such references.
 
     Returns:
         Dataset holding Arrow records read from the dataframes.
     """
+    import pandas as pd
+
+    if isinstance(dfs, pd.DataFrame):
```
Comment (clarkzinzow) [inline python/ray/data/read_api.py:529]: These functions actually take DataFrame refs, not pure DataFrames, so this should be:
```suggestion
    if isinstance(dfs, ray.ObjectRef):
```

File: python/ray/data/read_api.py [hunk 2]
```diff
@@ -546,17 +553,22 @@ def from_numpy(ndarrays: List[ObjectRef[np.ndarray]]) -> Dataset[ArrowRow]:
...
-                or its streaming format in bytes.
+        dfs: A Ray object references to Arrow tables or a list of such
+            references.
 
     Returns:
         Dataset holding Arrow records from the tables.
     """
+    import pyarrow as pa
+
+    if isinstance(tables, pa.Table):
```
Comment (clarkzinzow) [inline python/ray/data/read_api.py:569]: Same here:
```suggestion
    if isinstance(tables, ray.ObjectRef):
```

**Review by clarkzinzow — CHANGES_REQUESTED** [review]
My bad, I should have caught this on my first review!

**Comment by clarkzinzow** [comment]
@sighingnow We made a small update to these APIs, where `from_pandas()` and `from_arrow()` now take concrete dataframes and tables instead of object refs, while the old ref-based APIs are now `from_pandas_refs()` and `from_arrow_refs()`. Could you rebase onto/merge master?

Also, I think that a variadic API would be nicer:
```python
def from_pandas(*dfs: "pd.DataFrame") -> Dataset:
    ...
df = ...
ds = ray.data.from_pandas(df)

df2 = ...
dfs = [df, df2]
ds = ray.data.from_pandas(*dfs)
```

**Comment by sighingnow** [comment]
Hi @clarkzinzow any further comments on this pull request?

**Comment by sighingnow** [comment]
> Hi @clarkzinzow any further comments on this pull request?

It seems a deleted review comment. I found it in my email but cannot see it on github. Maybe it already looks good to land?

**Comment by clarkzinzow** [comment]
Hi @sighingnow, this looks nearly there, any chance you could switch to the variadic args API that I mentioned in the latter half of this comment? https://github.com/ray-project/ray/pull/18205#issuecomment-932639626

**Review by clarkzinzow — APPROVED** [review]
LGTM!

**Review by ericl — APPROVED** [review]

**Review by ericl — APPROVED** [review]

**Pull request merged**